coredump: add bits to enable dumping privilege stack during coredump

dcpleung opened the pull request:
See individual commits for details.

Only RISC-V, x86 and Xtensa are supported right now as they are simple enough to figure out where the privilege stacks are.

dcpleung committed:
arch: add interface to dump privileged stack in coredump

This adds a interface to allow coredump to dump privileged
stack which is defined in architecture specific way.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
debug: coredump: dump privileged stack

This adds the bits to call into architecture code to dump
the privileged stack for user threads.

The weak implementation is simply there as a stub until
all architectures have implemented the associated function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: debug/coredump: crash from user thread if userspace

Modify the test to spawn a user thread if it is running
with userspace enabled. This allows testing coredump when
the crashing thread is a user thread.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: coredump: support dumping privilege stack

Adds the bits to support dumping privilege stack during
coredump.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: coredump: support dumping privilege stack

Adds the bits to support dumping privilege stack during
coredump.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

agiardin commented on arch/Kconfig:
```diff
@@ -114,6 +115,7 @@ config RISCV
 	bool
 	select ARCH_IS_SET
 	select ARCH_SUPPORTS_COREDUMP
+	select ARCH_SUPPORTS_COREDUMP_PRIV_STACKS
```
`select ARCH_SUPPORTS_COREDUMP_PRIV_STACKS` needs to also be added to `config XTENSA` right?

dcpleung replied:
On Xtensa, coredump is only enabled by the SoC or board. So this is not done via arch kconfig.

dcpleung replied:
I have moved the kconfig to the SoC instead of board, which is more appropriate.

agiardin commented on subsys/debug/coredump/Kconfig:
```diff
@@ -107,4 +107,15 @@ config DEBUG_COREDUMP_THREADS_METADATA
+	help
+	  Dump the privilege stack of user threads.
+
+	  Say n to conserve space on coredump backend or that you will never
```
```suggestion
	  Say n to conserve space on coredump backend or if you will never
```

dcpleung replied:
Done.

dcpleung committed:
soc: cdns/dc233c: advertise coredump with privilege stack

This lets the SoC to select the correct kconfigs to show that
it supports coredump, and with the ability to dump privilege
stack.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

ycsin commented on arch/riscv/core/coredump.c:
```diff
@@ -116,3 +117,21 @@ uint16_t arch_coredump_tgt_code_get(void)
 }
+
+#if defined(CONFIG_DEBUG_COREDUMP_DUMP_THREAD_PRIV_STACK)
+void arch_coredump_priv_stack_dump(struct k_thread *thread)
```
There's a similar implementation in `stacktrace.c`

https://github.com/zephyrproject-rtos/zephyr/blob/74c5c84c66d59320ab39b58662d60e6cb6f188c2/arch/riscv/core/stacktrace.c#L55-L71

dcpleung replied:
Thanks. Updated.

dcpleung committed:
riscv: support dumping privilege stack during coredump

Adds some bits to enable dumping privilege stack during
coredump.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Twister failure is due to: https://github.com/zephyrproject-rtos/zephyr/issues/78677

mrkhldn commented on tests/subsys/debug/coredump/testcase.yaml:
```diff
@@ -22,3 +22,31 @@ tests:
         - "E: #CD:4([dD])([0-9a-fA-F]+)"
         - "E: #CD:END#"
         - "k_sys_fatal_error_handler"
+  debug.coredump.logging_backend.usespace:
```
debug.coredump.logging_backend.userspace

dcpleung replied:
Oops... fixed.

mrkhldn commented on tests/subsys/debug/coredump/testcase.yaml:
```diff
@@ -22,3 +22,31 @@ tests:
+        - "E: #CD:5([aA])45([0-9a-fA-F]+)"
+        - "E: #CD:41([0-9a-fA-F]+)"
+        - "E: #CD:4([dD])([0-9a-fA-F]+)"
+        - "E: #CD:4([dD])([0-9a-fA-F]+)"
```
Should there be at least one more line of the following from the call to arch_coredump_priv_stack_dump

"E: #CD:4([dD])([0-9a-fA-F]+)"

dcpleung replied:
Yes... added.
